sway/commands/bind.c

```diff
@@ -649,101 +646,144 @@ void seat_execute_command(struct sway_seat *seat, struct sway_binding *binding)
+	int total_combinations = 1;
+	for (int i = 0; i < n; ++i) {
+		total_combinations *= sets[i]->length;
+	}
+		
+	// Allocate memory for the current_result
+	int result_size = 0;
+	xkb_keycode_t **result = malloc(total_combinations * sizeof(xkb_keycode_t*));
+	xkb_keycode_t *current_product = malloc(n * sizeof(xkb_keycode_t));
+	cartesian_product_helper(sets, n, result, &result_size, current_product, 0);
+	free(current_product);
+
+	return result;
 }
 
+/*
+ * Convert keysyms to keycodes for --to-code. If any keysym does not have at
+ * least 1 keycode: halt conversion and unset BINDING_CODE. Assumes
+ * identify_key() read in binding->keys in keysym format vs keycode format
+ */
 bool translate_binding(struct sway_binding *binding) {
-	if ((binding->flags & BINDING_CODE) == 0) {
-		return true;
-	}
 
-	switch (binding->type) {
-	// a bindsym to translate
-	case BINDING_KEYSYM:
-		binding->syms = binding->keys;
-		binding->keys = create_list();
-		break;
-	// a bindsym to re-translate
-	case BINDING_KEYCODE:
-		list_free_items_and_destroy(binding->keys);
-		binding->keys = create_list();
-		break;
-	default:
+	if (binding->type != BINDING_KEYSYM ||
+		(binding->flags & BINDING_CODE) == 0) {
 		return true;
 	}
 
-	for (int i = 0; i < binding->syms->length; ++i) {
-		xkb_keysym_t *keysym = binding->syms->items[i];
-		struct keycode_matches matches = get_keycode_for_keysym(*keysym);
+	int num_syms = binding->keys->length;
+	list_t **sym2code = malloc(num_syms * sizeof(list_t*)); 
+	// Collect all keycodes for all keysyms
+	for (int i = 0; i < num_syms; i++) {
+		struct keycode_matches matches = {
+			.keysym = (xkb_keysym_t*)binding->keys->items[i],
+			.keycodes = create_list(),
+			.error = false
+		};
 
-		if (matches.count != 1) {
-			sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
-					" a single keycode (found %d matches)",
-					*keysym, matches.count);
+		xkb_keymap_key_for_each(
+				xkb_state_get_keymap(config->keysym_translation_state),
+				add_matching_keycodes, &matches);
+
+		if (matches.error) {
+			sway_log(SWAY_ERROR, "Failed to allocate memory for keycodes while iterating for keysym %" PRIu32, *matches.keysym);
 			goto error;
 		}
 
-		xkb_keycode_t *keycode = malloc(sizeof(xkb_keycode_t));
-		if (!keycode) {
-			sway_log(SWAY_ERROR, "Unable to allocate memory for a keycode");
-			goto error;
+		if (matches.keycodes->length == 0) {
+		   sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
+					  "any keycodes", *matches.keysym);
+		   goto error;
 		}
 
-		*keycode = matches.keycode;
-		list_add(binding->keys, keycode);
+		sym2code[i] = matches.keycodes;
 	}
 
-	list_qsort(binding->keys, key_qsort_cmp);
-	binding->type = BINDING_KEYCODE;
+	// If any keycode maps to more than one keysym, use all combinations.
+	xkb_keycode_t **combos = cartesian_product(sym2code, num_syms);
+
+	for (int i = 0; i < num_syms; i++) {
+		struct sway_binding * copy_binding = malloc(sizeof(struct sway_binding));
```

Comment:
Three nits:
1. Should be `*copy_binding`, not `* copy_binding`. Same goes if there are more asterisks.
2. Our preferred style is to sizeof the variable itself ,so `sizeof(*copy_binding)` rather than a named type. It's not unusual that one ends up accidentally typing the wrong type, but typing the wrong name tends to be easier to catch.
3. NULL check

Applies elsewhere too, especially the first one.

Reply:
1. Agreed, oops
2. I can definitely do that. Every other sizeof call in `bind.c` does it by type though. Should I change those?  
![image](https://github.com/user-attachments/assets/c20433bd-93d4-4e00-8fcf-fe5beb9e3539)
![image](https://github.com/user-attachments/assets/0dd05baa-5d9e-41e4-b6d5-589601cb6480)
![image](https://github.com/user-attachments/assets/1e7dee78-b916-455a-9e76-40dde625cdf6)
![image](https://github.com/user-attachments/assets/5210ff4d-8659-4aed-aeb1-a0c0d5e76d39)
3. Adding.

Reply:
Yeah, we went through the paces of updating all sizeofs in wlroots (https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4367), but we haven't done such an effort in sway. :/

> I can definitely do that. Every other sizeof call in bind.c does it by type though. Should I change those

No, unless the PR is specifically a tree-wide cleanup, stick to modifying the code you need to change. Otherwise you end up with a diff where it's hard to tell format changes from logic changes.

> Actually some of those examples above are the first invocation to the specified type, so they should refer to types anyway.

The first invocation of the specified type is the declaration, and you can already reference the name in the right-hand side of the declaring assignment. The following is perfectly valid and is the preferred construct:

```c
struct sway_binding *binding = calloc(1, sizeof(*binding));
```